File: baekjoon/src/tags/dfsAndBfs/No1012.java
package tags.dfsAndBfs;

import java.io.*;
import java.util.StringTokenizer;

/**
 * https://www.acmicpc.net/problem/1012
 */
public class No1012 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    public static int width;
    public static int height;
    public static int[][] graph;

    public static void setGraph() throws IOException {
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        width = Integer.parseInt(tokenizer.nextToken());
        height = Integer.parseInt(tokenizer.nextToken());
        graph = new int[height][width];

        int input = Integer.parseInt(tokenizer.nextToken());
        for (int i = 0; i < input; i++) {
            tokenizer = new StringTokenizer(reader.readLine());
            int col = Integer.parseInt(tokenizer.nextToken());
            int row = Integer.parseInt(tokenizer.nextToken());
            graph[row][col] = 1;
        }
    }

    public static void bfs(int row, int col) {
        graph[row][col] = -1;
        if (row - 1 >= 0 && graph[row - 1][col] == 1)
            bfs(row - 1, col);
        if (col + 1 < width && graph[row][col + 1] == 1)
            bfs(row, col + 1);
        if (row + 1 < height && graph[row + 1][col] == 1)
            bfs(row + 1, col);
        if (col - 1 >= 0 && graph[row][col - 1] == 1)
            bfs(row, col - 1);
    }

    public static void main(String[] args) throws IOException {
        int testCase = Integer.parseInt(reader.readLine());
        for (int i = 0; i < testCase; i++) {
            setGraph();
            int worm = 0;
            for (int j = 0; j < height; j++) {
                for (int k = 0; k < width; k++) {
                    if (graph[j][k] == 1) {
                        bfs(j, k);
                        worm++;
                    }
                }
            }
            writer.write(String.valueOf(worm));
            writer.newLine();
        }
        writer.flush();
    }
}

File: baekjoon-classification/src/dfs/No1012.kt
package dfs

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

/**
 * https://www.acmicpc.net/problem/1012
 */
private val reader = BufferedReader(InputStreamReader(System.`in`))
private val writer = BufferedWriter(OutputStreamWriter(System.out))

private var height: Int = 0
private var width: Int = 0
private lateinit var graph: Array<IntArray>

private val dx = intArrayOf(1, 0, -1, 0)
private val dy = intArrayOf(0, -1, 0, 1)

fun main() {
    val testCase = reader.readLine().toInt()
    repeat(testCase) {
        var count = 0

        var tokenizer = StringTokenizer(reader.readLine())
        height = tokenizer.nextToken().toInt()
        width = tokenizer.nextToken().toInt()
        val numberOfInput = tokenizer.nextToken().toInt()

        graph = Array(height) { IntArray(width) }
        repeat(numberOfInput) {
            tokenizer = StringTokenizer(reader.readLine())
            val row = tokenizer.nextToken().toInt()
            val col = tokenizer.nextToken().toInt()
            graph[row][col] = 1
        }

        for (i in 0 until height) {
            for (j in 0 until width) {
                if (graph[i][j] == 1) {
                    dfs(i, j)
                    count++
                }
            }
        }
        writer.write("$count")
        writer.newLine()
    }
    writer.flush()
}

private fun dfs(row: Int, col: Int) {
    graph[row][col] = -1
    for (i in 0..3) {
        val nextRow = row + dy[i]
        val nextCol = col + dx[i]
        if (isInBoundary(nextRow, nextCol) && graph[nextRow][nextCol] == 1)
            dfs(nextRow, nextCol)
    }
}

private fun isInBoundary(row: Int, col: Int) = row in 0 until height && col in 0 until width